esx_service/utils/auth_data.py

```diff
@@ -0,0 +1,306 @@
+    def set_datastore_access_privileges(self, conn, privileges):
+        # If we want to change privileges for <tenant1, ds1> and <tenant2, ds2>
+        # Do we call this function one time or two times?
+        tenant_id = self.id
+        for p in privileges:
+            p['tenant_id'] = tenant_id
+                            
+        conn.executemany(
+            """
+            insert or ignore into privileges values
+            (:tenant_id, :datastore, :global_visibility, :create_volume,
+            :delete_volume, :mount_volume, :max_volume_size, :usage_quota)
+            """,
+            privileges
+        )
+        
+        for p in privileges:
+            p['tenant_id'] = tenant_id
+            column_list = ['tenant_id', 'datastore', 'global_visibility', 'create_volume',
+                            'delete_volume', 'mount_volume', 'max_volume_size', 'usage_quota']
+            update_list = []
+            for col in column_list:
+                update_list.append(p[col])
+            update_list.append(tenant_id)
+            update_list.append(p['datastore'])    
+            
+            #print update_list
+
+            conn.execute(
+                """
+                update or ignore privileges set
+                tenant_id =?, 
+                datastore=?, 
+                global_visibility=?,
+                create_volume=?,
+                delete_volume=?,
+                mount_volume=?,
+                max_volume_size=?,
+                usage_quota=?
+                where tenant_id=? and datastore=?
+                """,
+                update_list
+            )
+        conn.commit()
+
+class AuthorizationDataManager:
+    """
+    This class abstracts the creation, modification and retrieval of
+    authorization data used by vmdk_ops as well as the VMODL interface for
+    Docker volume management.
+    """
+
+    def __init__(self, db_path):
+        self.db_path = db_path
+        self.conn = None
+
+    def __del__(self):
+        if self.conn:
+            self.conn.close()
+
+    def connect(self):
+        """
+        Connect to a sqlite database file given by `db_path`. Ensure foreign key
+        constraints are enabled on the database and set the return type for
+        select operations to dict-like 'Rows' instead of tuples.
+
+        Raises a ConnectionFailed exception of
+        """
+        self.conn = sqlite3.connect(self.db_path)
+
+        # Return rows as Row instances instead of tuples
+        self.conn.row_factory = sqlite3.Row
+
+        if not self.conn:
+            raise DbConnectionError(self.db_path)
+
+    def create_tables(self):
+        """
+        Create the 3 tables used for per-datastore authorization.  This function
+        should only be called once per datastore.  It will raise an exception if
+        the schema file isn't accessible or the tables already exist.
+        """
+        
+        with open('./docker-volume-auth-schema.sql') as f:
```

Comment:
This code is in Andy's original implementation. My understanding is that he choose to put the schema in a file is to make the code looks much cleaner?  @andrewjstone correct me if I misunderstand anything.